File: src/harmonic/sacredEggs.ts
/**
 * @file sacredEggs.ts
 * @module harmonic/sacredEggs
 * @layer Layer 5, Layer 7, Layer 8, Layer 13
 * @component Sacred Eggs — Ritual-Based Conditional Secret Distribution
 * @version 3.0.0
 * @since 2026-02-07
 *
 * Predicate-gated AEAD encryption where decryption requires ALL four
 * predicates to be satisfied simultaneously:
 *
 *   P1(tongue):   Correct Sacred Tongue identity
 *   P2(geometry): Correct position in Poincaré ball (L5 metric)
 *   P3(path):     Valid PHDM Hamiltonian path history
 *   P4(quorum):   k-of-n threshold met
 *
 * If ANY predicate fails, the derived key is wrong and AEAD decryption
 * produces an auth failure — no information leaks about WHICH predicate
 * was wrong (fail-to-noise property).
 *
 * Validated by experiments SE-1 (16-case predicate matrix),
 * SE-2 (output collapse), SE-3 (geometry key separation).
 */

import * as crypto from 'crypto';
import { hkdfSha256 } from '../crypto/hkdf.js';
import { hyperbolicDistance, projectToBall } from './hyperbolic.js';
import type { TongueCode } from './sacredTongues.js';

// ═══════════════════════════════════════════════════════════════
// Types
// ═══════════════════════════════════════════════════════════════

/** Phase angles for each Sacred Tongue (60° spacing) */
export const EGG_TONGUE_PHASES: Record<TongueCode, number> = {
  ko: 0,                    // 0°
  av: Math.PI / 3,          // 60°
  ru: (2 * Math.PI) / 3,    // 120°
  ca: Math.PI,              // 180°
  um: (4 * Math.PI) / 3,    // 240°
  dr: (5 * Math.PI) / 3,    // 300°
};

/** A single party's share in the quorum */
export interface QuorumShare {
  partyId: number;
  share: Buffer;  // 32-byte secret share
}

/** Configuration for a Sacred Egg */
export interface SacredEggConfig {
  /** Sacred Tongue code for this egg */
  tongue: TongueCode;
  /** Expected position in Poincaré ball (6D) */
  geometryCenter: number[];
  /** Max hyperbolic distance for geometry match (default 0.5) */
  geometryThreshold?: number;
  /** Quorum threshold (k of n) */
  quorumK: number;
  /** Total quorum parties */
  quorumN: number;
}

/** A sealed Sacred Egg */
export interface SacredEgg {
  /** Encrypted secret (nonce || ciphertext || tag) */
  ciphertext: Buffer;
  /** Authenticated additional data */
  aad: Buffer;
  /** Sacred Tongue code */
  tongue: TongueCode;
  /** Expected geometry center */
  geometryCenter: number[];
  /** Geometry match threshold */
  geometryThreshold: number;
  /** Hash of expected PHDM path */
  pathCommitment: Buffer;
  /** Quorum threshold */
  quorumK: number;
  /** Total quorum parties */
  quorumN: number;
  /** Salt for key derivation */
  salt: Buffer;
}

/** Result of an unseal attempt */
export interface UnsealResult {
  success: boolean;
  secret?: Buffer;
}

// ═══════════════════════════════════════════════════════════════
// Quorum Operations
// ═══════════════════════════════════════════════════════════════

/**
 * Generate quorum shares for n parties with threshold k.
 *
 * Uses deterministic derivation from a seed for reproducibility.
 * In production, use Shamir's Secret Sharing.
 */
export function generateQuorum(
  n: number,
  k: number,
  seed: Buffer
): { shares: QuorumShare[]; combined: Buffer } {
  const shares: QuorumShare[] = [];
  for (let i = 0; i < n; i++) {
    const shareData = Buffer.alloc(4);
    shareData.writeUInt32LE(i);
    const share = crypto
      .createHash('sha256')
      .update(Buffer.concat([seed, shareData]))
      .digest();
    shares.push({ partyId: i, share: Buffer.from(share) });
  }
  const combined = crypto
    .createHash('sha256')
    .update(Buffer.concat([seed, Buffer.from(':quorum-master')]))
    .digest();
  return { shares, combined };
}

/**
 * Combine k shares into quorum material.
 *
 * If fewer than k shares provided, produces wrong material (fail-to-noise).
 */
export function combineShares(shares: QuorumShare[], k: number): Buffer {
  if (shares.length < k) {
    return Buffer.alloc(32); // insufficient — will produce wrong key
  }
  const sorted = [...shares].sort((a, b) => a.partyId - b.partyId).slice(0, k);
  let combined = Buffer.alloc(32);
  for (const s of sorted) {
    combined = Buffer.from(combined.map((b, i) => b ^ s.share[i]));
  }
  return crypto
    .createHash('sha256')
    .update(Buffer.concat([combined, Buffer.from(':quorum-combined')]))
    .digest();
}

// ═══════════════════════════════════════════════════════════════
// Path Hashing
// ═══════════════════════════════════════════════════════════════

/**
 * Hash a PHDM path (sequence of polyhedron indices).
 */
export function pathHash(pathIndices: number[]): Buffer {
  const data = 'phdm:path:' + pathIndices.join(',');
  return crypto.createHash('sha256').update(data).digest();
}

// ═══════════════════════════════════════════════════════════════
// Key Derivation
// ═══════════════════════════════════════════════════════════════

/**
 * Derive the AEAD key from all four predicates.
 *
 * Each predicate contributes independent key material.
 * The final key is derived from their concatenation via HKDF.
 * If ANY input differs, the derived key is completely different (avalanche).
 */
export function deriveEggKey(
  tongue: TongueCode,
  geometryPoint: number[],
  pathIndices: number[],
  quorumMaterial: Buffer,
  salt: Buffer
): Buffer {
  // P1: Tongue material
  const tonguePhase = EGG_TONGUE_PHASES[tongue] ?? 0;
  const tonguePhaseBytes = Buffer.alloc(8);
  tonguePhaseBytes.writeDoubleBE(tonguePhase);
  const tongueMaterial = crypto
    .createHash('sha256')
    .update(
      Buffer.concat([
        Buffer.from('sacred-egg:tongue:'),
        Buffer.from(tongue),
        tonguePhaseBytes,
      ])
    )
    .digest();

  // P2: Geometry material — point coordinates at full precision
  const geoFloats = Buffer.alloc(geometryPoint.length * 8);
  for (let i = 0; i < geometryPoint.length; i++) {
    geoFloats.writeDoubleBE(geometryPoint[i], i * 8);
  }
  const geometryMaterial = crypto
    .createHash('sha256')
    .update(Buffer.concat([Buffer.from('sacred-egg:geometry:'), geoFloats]))
    .digest();

  // P3: Path material
  const pathMaterial = pathHash(pathIndices);

  // P4: Quorum material
  const quorumMat = crypto
    .createHash('sha256')
    .update(Buffer.concat([Buffer.from('sacred-egg:quorum:'), quorumMaterial]))
    .digest();

  // Combine all four into AEAD key
  const ikm = Buffer.concat([tongueMaterial, geometryMaterial, pathMaterial, quorumMat]);
  const info = Buffer.from('sacred-egg:aead-key:v1');
  return hkdfSha256(ikm, salt, info, 32);
}

// ═══════════════════════════════════════════════════════════════
// AEAD (AES-256-GCM)
// ═══════════════════════════════════════════════════════════════

/**
 * Encrypt with AES-256-GCM AEAD.
 *
 * @returns Buffer containing: nonce(12) || ciphertext || tag(16)
 */
function aeadEncrypt(key: Buffer, plaintext: Buffer, aad: Buffer): Buffer {
  const nonce = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv('aes-256-gcm', key, nonce);
  cipher.setAAD(aad);
  const ct = Buffer.concat([cipher.update(plaintext), cipher.final()]);
  const tag = cipher.getAuthTag();
  return Buffer.concat([nonce, ct, tag]);
}

/**
 * Decrypt with AES-256-GCM AEAD.
 *
 * @returns Plaintext on success, null on auth failure (fail-to-noise)
 */
function aeadDecrypt(key: Buffer, ciphertext: Buffer, aad: Buffer): Buffer | null {
  if (ciphertext.length < 28) return null; // nonce(12) + tag(16) minimum

  const nonce = ciphertext.subarray(0, 12);
  const tag = ciphertext.subarray(ciphertext.length - 16);
  const ct = ciphertext.subarray(12, ciphertext.length - 16);

  try {
    const decipher = crypto.createDecipheriv('aes-256-gcm', key, nonce);
    decipher.setAAD(aad);
    decipher.setAuthTag(tag);
    return Buffer.concat([decipher.update(ct), decipher.final()]);
  } catch {
    return null; // Auth failed — fail-to-noise
  }
}

// ═══════════════════════════════════════════════════════════════
// Sacred Egg Operations
// ═══════════════════════════════════════════════════════════════

/**
 * Seal a secret into a Sacred Egg.
 *
 * The secret will only be recoverable if ALL four predicates
 * are presented correctly during unseal.
 *
 * @param secret - The secret to protect
 * @param config - Egg configuration (tongue, geometry, quorum params)
 * @param geometryPoint - Sealer's position in Poincaré ball
 * @param pathIndices - PHDM Hamiltonian path history
 * @param quorumShares - Available quorum shares (must have >= k)
 * @returns Sealed Sacred Egg
 */
export function sealEgg(
  secret: Buffer,
  config: SacredEggConfig,
  geometryPoint: number[],
  pathIndices: number[],
  quorumShares: QuorumShare[]
): SacredEgg {
  const salt = crypto.randomBytes(32);
  const quorumMaterial = combineShares(quorumShares, config.quorumK);

  const key = deriveEggKey(config.tongue, geometryPoint, pathIndices, quorumMaterial, salt);

  const aad = Buffer.from(
    JSON.stringify({
      type: 'sacred-egg',
      version: 'v1',
      tongue: config.tongue,
      quorum_k: config.quorumK,
      quorum_n: config.quorumN,
    })
  );

  const ciphertext = aeadEncrypt(key, secret, aad);

  return {
    ciphertext,
    aad,
    tongue: config.tongue,
    geometryCenter: [...geometryPoint],
    geometryThreshold: config.geometryThreshold ?? 0.5,
    pathCommitment: pathHash(pathIndices),
    quorumK: config.quorumK,
    quorumN: config.quorumN,
    salt,
  };
}

/**
 * Attempt to unseal a Sacred Egg.
 *
 * ALL four predicates must match for decryption to succeed.
 * On failure, returns { success: false } with no information
 * about which predicate was wrong (fail-to-noise).
 *
 * @param egg - The Sacred Egg to unseal
 * @param tongue - Claimed tongue code
 * @param geometryPoint - Claimer's position in Poincaré ball
 * @param pathIndices - Claimed PHDM path history
 * @param quorumShares - Available quorum shares
 * @returns UnsealResult with secret on success, nothing on failure
 */
export function unsealEgg(
  egg: SacredEgg,
  tongue: TongueCode,
  geometryPoint: number[],
  pathIndices: number[],
  quorumShares: QuorumShare[]
): UnsealResult {
  const quorumMaterial = combineShares(quorumShares, egg.quorumK);

  const key = deriveEggKey(tongue, geometryPoint, pathIndices, quorumMaterial, egg.salt);

  const result = aeadDecrypt(key, egg.ciphertext, egg.aad);

  if (result !== null) {
    return { success: true, secret: result };
  }
  return { success: false };
}

/**
 * Check if a geometry point is within the egg's threshold.
 *
 * This is an OPTIONAL pre-check — the real gating happens in key derivation.
 * Use this for UX (show distance feedback) but NEVER skip AEAD verification.
 */
export function checkGeometryProximity(
  egg: SacredEgg,
  candidatePoint: number[]
): { withinThreshold: boolean; distance: number } {
  const a = projectToBall(egg.geometryCenter);
  const b = projectToBall(candidatePoint);
  const distance = hyperbolicDistance(a, b);
  return {
    withinThreshold: distance <= egg.geometryThreshold,
    distance,
  };
}

File: src/symphonic_cipher/scbe_aethermoore/sacred_eggs.py
"""Sacred Eggs — Ritual-Based Conditional Secret Distribution (Python Reference)

Predicate-gated AEAD encryption where decryption requires ALL four
predicates to be satisfied simultaneously:

  P1(tongue):   Correct Sacred Tongue identity
  P2(geometry): Correct position in Poincaré ball
  P3(path):     Valid PHDM Hamiltonian path history
  P4(quorum):   k-of-n threshold met

If ANY predicate fails, the derived key is wrong and AEAD decryption
produces auth failure — no information leaks about WHICH predicate
was wrong (fail-to-noise).

Validated: SE-1 (predicate matrix), SE-2 (output collapse), SE-3 (geometry separation).
"""

import hashlib
import hmac
import json
import math
import os
import struct
from dataclasses import dataclass
from typing import List, Optional, Tuple

import numpy as np

# ═══════════════════════════════════════════════════════════════
# Constants
# ═══════════════════════════════════════════════════════════════

TONGUE_CODES = ('ko', 'av', 'ru', 'ca', 'um', 'dr')

TONGUE_PHASES = {
    'ko': 0.0,
    'av': math.pi / 3,
    'ru': 2 * math.pi / 3,
    'ca': math.pi,
    'um': 4 * math.pi / 3,
    'dr': 5 * math.pi / 3,
}

EPSILON = 1e-10

# ═══════════════════════════════════════════════════════════════
# Crypto primitives
# ═══════════════════════════════════════════════════════════════


def hkdf_sha256(ikm: bytes, salt: bytes, info: bytes, length: int = 32) -> bytes:
    """HKDF-SHA256 (RFC 5869)."""
    if not salt:
        salt = b'\x00' * 32
    prk = hmac.new(salt, ikm, hashlib.sha256).digest()
    t = b''
    okm = b''
    counter = 1
    while len(okm) < length:
        t = hmac.new(prk, t + info + bytes([counter]), hashlib.sha256).digest()
        okm += t
        counter += 1
    return okm[:length]


def aead_encrypt(key: bytes, plaintext: bytes, aad: bytes) -> bytes:
    """Encrypt-then-HMAC AEAD. Format: nonce(16) || ct || mac(32)."""
    nonce = os.urandom(16)
    k_enc = hkdf_sha256(key, nonce, b'sacred-egg:enc', 32)
    k_mac = hkdf_sha256(key, nonce, b'sacred-egg:mac', 32)

    ct = bytearray(len(plaintext))
    for i in range(0, len(plaintext), 32):
        block_idx = i // 32
        ks = hashlib.sha256(k_enc + struct.pack('<Q', block_idx)).digest()
        for j in range(min(32, len(plaintext) - i)):
            ct[i + j] = plaintext[i + j] ^ ks[j]

    mac = hmac.new(k_mac, aad + nonce + bytes(ct), hashlib.sha256).digest()
    return nonce + bytes(ct) + mac


def aead_decrypt(key: bytes, ciphertext: bytes, aad: bytes) -> Optional[bytes]:
    """Decrypt and verify. Returns None on auth failure."""
    if len(ciphertext) < 48:
        return None

    nonce = ciphertext[:16]
    mac_received = ciphertext[-32:]
    ct = ciphertext[16:-32]

    k_enc = hkdf_sha256(key, nonce, b'sacred-egg:enc', 32)
    k_mac = hkdf_sha256(key, nonce, b'sacred-egg:mac', 32)

    mac_expected = hmac.new(k_mac, aad + nonce + ct, hashlib.sha256).digest()
    if not hmac.compare_digest(mac_received, mac_expected):
        return None

    pt = bytearray(len(ct))
    for i in range(0, len(ct), 32):
        block_idx = i // 32
        ks = hashlib.sha256(k_enc + struct.pack('<Q', block_idx)).digest()
        for j in range(min(32, len(ct) - i)):
            pt[i + j] = ct[i + j] ^ ks[j]

    return bytes(pt)


# ═══════════════════════════════════════════════════════════════
# Geometry
# ═══════════════════════════════════════════════════════════════


def poincare_distance(u: np.ndarray, v: np.ndarray) -> float:
    """Hyperbolic distance in Poincaré ball."""
    diff_sq = float(np.sum((u - v) ** 2))
    u_sq = float(np.sum(u ** 2))
    v_sq = float(np.sum(v ** 2))
    u_factor = max(EPSILON, 1.0 - u_sq)
    v_factor = max(EPSILON, 1.0 - v_sq)
    arg = 1.0 + 2.0 * diff_sq / (u_factor * v_factor)
    return float(np.arccosh(max(1.0, arg)))


def project_to_ball(p: np.ndarray, max_norm: float = 1.0 - 1e-10) -> np.ndarray:
    """Project point into Poincaré ball."""
    n = float(np.linalg.norm(p))
    if n < max_norm:
        return p.copy()
    return p * (max_norm / n)


# ═══════════════════════════════════════════════════════════════
# Quorum
# ═══════════════════════════════════════════════════════════════


@dataclass
class QuorumShare:
    """A party's share in the quorum."""
    party_id: int
    share: bytes


def generate_quorum(n: int, k: int, seed: bytes) -> Tuple[List[QuorumShare], bytes]:
    """Generate n shares with threshold k."""
    shares = []
    for i in range(n):
        share = hashlib.sha256(seed + struct.pack('<I', i)).digest()
        shares.append(QuorumShare(party_id=i, share=share))
    combined = hashlib.sha256(seed + b':quorum-master').digest()
    return shares, combined


def combine_shares(shares: List[QuorumShare], k: int) -> bytes:
    """Combine k shares into quorum material."""
    if len(shares) < k:
        return b'\x00' * 32
    sorted_shares = sorted(shares, key=lambda s: s.party_id)[:k]
    combined = b'\x00' * 32
    for s in sorted_shares:
        combined = bytes(a ^ b for a, b in zip(combined, s.share))
    return hashlib.sha256(combined + b':quorum-combined').digest()


# ═══════════════════════════════════════════════════════════════
# Path hashing
# ═══════════════════════════════════════════════════════════════


def path_hash(path_indices: List[int]) -> bytes:
    """Hash a PHDM path."""
    data = b'phdm:path:' + b','.join(str(i).encode() for i in path_indices)
    return hashlib.sha256(data).digest()


# ═══════════════════════════════════════════════════════════════
# Key Derivation
# ═══════════════════════════════════════════════════════════════


def derive_egg_key(
    tongue_code: str,
    geometry_point: np.ndarray,
    path_indices: List[int],
    quorum_material: bytes,
    salt: bytes,
) -> bytes:
    """Derive AEAD key from all four predicates."""
    tongue_phase = TONGUE_PHASES.get(tongue_code, 0.0)
    tongue_material = hashlib.sha256(
        b'sacred-egg:tongue:' + tongue_code.encode() +
        struct.pack('<d', tongue_phase)
    ).digest()

    geo_bytes = geometry_point.tobytes()
    geometry_material = hashlib.sha256(
        b'sacred-egg:geometry:' + geo_bytes
    ).digest()

    path_material = path_hash(path_indices)

    quorum_mat = hashlib.sha256(
        b'sacred-egg:quorum:' + quorum_material
    ).digest()

    ikm = tongue_material + geometry_material + path_material + quorum_mat
    return hkdf_sha256(ikm, salt, b'sacred-egg:aead-key:v1', 32)


# ═══════════════════════════════════════════════════════════════
# Sacred Egg
# ═══════════════════════════════════════════════════════════════


@dataclass
class SacredEgg:
    """A sealed Sacred Egg."""
    ciphertext: bytes
    aad: bytes
    tongue_code: str
    geometry_center: np.ndarray
    geometry_threshold: float
    path_commitment: bytes
    quorum_k: int
    quorum_n: int
    salt: bytes


def seal_egg(
    secret: bytes,
    tongue_code: str,
    geometry_point: np.ndarray,
    path_indices: List[int],
    quorum_shares: List[QuorumShare],
    quorum_k: int,
    quorum_n: int,
) -> SacredEgg:
    """Seal a secret into a Sacred Egg."""
    salt = os.urandom(32)
    quorum_material = combine_shares(quorum_shares, quorum_k)
    key = derive_egg_key(tongue_code, geometry_point, path_indices, quorum_material, salt)

    aad = json.dumps({
        'type': 'sacred-egg',
        'version': 'v1',
        'tongue': tongue_code,
        'quorum_k': quorum_k,
        'quorum_n': quorum_n,
    }, sort_keys=True).encode()

    ciphertext = aead_encrypt(key, secret, aad)

    return SacredEgg(
        ciphertext=ciphertext,
        aad=aad,
        tongue_code=tongue_code,
        geometry_center=geometry_point.copy(),
        geometry_threshold=0.5,
        path_commitment=path_hash(path_indices),
        quorum_k=quorum_k,
        quorum_n=quorum_n,
        salt=salt,
    )


def unseal_egg(
    egg: SacredEgg,
    tongue_code: str,
    geometry_point: np.ndarray,
    path_indices: List[int],
    quorum_shares: List[QuorumShare],
) -> Optional[bytes]:
    """Attempt to unseal. Returns secret on success, None on failure."""
    quorum_material = combine_shares(quorum_shares, egg.quorum_k)
    key = derive_egg_key(tongue_code, geometry_point, path_indices, quorum_material, egg.salt)
    return aead_decrypt(key, egg.ciphertext, egg.aad)


def check_geometry_proximity(
    egg: SacredEgg, candidate: np.ndarray
) -> Tuple[bool, float]:
    """Optional pre-check for geometry proximity."""
    a = project_to_ball(egg.geometry_center)
    b = project_to_ball(candidate)
    d = poincare_distance(a, b)
    return d <= egg.geometry_threshold, d


# ═══════════════════════════════════════════════════════════════
# Self-tests
# ═══════════════════════════════════════════════════════════════

def self_test():
    """Run basic validation tests."""
    print("Sacred Eggs Python reference — self-tests")
    passed = 0
    total = 0

    # Test 1: Seal/unseal with correct predicates
    total += 1
    pt = project_to_ball(np.array([0.1, 0.2, 0.3, -0.1, 0.05, -0.15]))
    path = list(range(16))
    seed = os.urandom(32)
    shares, _ = generate_quorum(5, 3, seed)
    secret = b'test secret 12345678901234567890'
    egg = seal_egg(secret, 'ko', pt, path, shares, 3, 5)
    result = unseal_egg(egg, 'ko', pt, path, shares[:3])
    assert result == secret, f"Correct predicates failed: {result}"
    passed += 1
    print(f"  [PASS] Correct predicates → decrypt")

    # Test 2: Wrong tongue fails
    total += 1
    result = unseal_egg(egg, 'dr', pt, path, shares[:3])
    assert result is None, "Wrong tongue should fail"
    passed += 1
    print(f"  [PASS] Wrong tongue → None")

    # Test 3: Wrong geometry fails
    total += 1
    wrong_pt = project_to_ball(np.array([0.5, 0.5, 0.5, 0.5, 0.5, 0.5]))
    result = unseal_egg(egg, 'ko', wrong_pt, path, shares[:3])
    assert result is None, "Wrong geometry should fail"
    passed += 1
    print(f"  [PASS] Wrong geometry → None")

    # Test 4: Wrong path fails
    total += 1
    wrong_path = list(range(15, -1, -1))
    result = unseal_egg(egg, 'ko', pt, wrong_path, shares[:3])
    assert result is None, "Wrong path should fail"
    passed += 1
    print(f"  [PASS] Wrong path → None")

    # Test 5: Wrong quorum fails
    total += 1
    wrong_shares = [QuorumShare(i, os.urandom(32)) for i in range(3)]
    result = unseal_egg(egg, 'ko', pt, path, wrong_shares)
    assert result is None, "Wrong quorum should fail"
    passed += 1
    print(f"  [PASS] Wrong quorum → None")

    # Test 6: Geometry proximity check
    total += 1
    within, dist = check_geometry_proximity(egg, pt)
    assert within and dist < 0.001, f"Same point should be within threshold: d={dist}"
    passed += 1
    print(f"  [PASS] Geometry proximity check")

    print(f"\n  {passed}/{total} tests passed")
    return passed == total


if __name__ == '__main__':
    ok = self_test()
    exit(0 if ok else 1)
